### ISSUE TITLE
Relaxed concatenation for "diagonal"

### ISSUE THREAD

**Issue by FrozenPie1** [pola-rs/polars#9412]
### Problem description

One and a half years ago i asked if it was possibly to implement diagonal concatenation with auto cast of not matching dtypes of the same columns in different DataFrames. The statement from that time was "Auto-Casting of identical columns with different dtypes shouldn't exist" but diagonal stacking was still implemented which i'm still thankful for.
With https://github.com/pola-rs/polars/pull/9382 now in place (relaxed concatenation including auto-cast) there seems to have been a change of mind so i would like to renew my question if it would now be possible to implement relaxed concatenation option for diagonal concatenation.
The details with explanations can be found in the old issue https://github.com/pola-rs/polars/issues/1665.

**Comment by cmdlineluser** [comment]
The duckdb `read_*` methods have a `union_by_name=true` which does this, it's a handy feature.
```python
import duckdb
import fsspec
from   textwrap import dedent

csvs = [
    """
    first,last,age
    Aaaaa,Fffff,40
    Mmmm,Cccccc,59
    Hhhhh,Mmmmmm,39
    Cccccc,Rrrrrr,42
    Sssss,Niiiii,32
    """
    ,
    """
    first,last,city,postal
    Cccccc,Ssssssss,Pppppp,156846
    Ffffff,Ssss,Ssssss,684321
    Aa,Mmmmmm,Ggggg,158431
    Bbbbbb,Ccccccc,Kkkkkkk,543154
    Bbbbbb,Mmmmmmmmm,Rrrrrrr,354849
    """
    ,
    """
    first,last,age,phone,postal
    Rrr,Rr,old,(123) 456-4913,C1O 2I2
    Ggggg,Ppppp,young,(456) 789-9433,J8Z 7S8
    Ffffff,Gggg,young,(789) 321-2187,N7A 5N4
    Iii,Ppppp,mid,(123) 321-4999,N6H 3J8
    kkkkk,Dddddd,old,(456) 654-1427,R0I 4V1
    """
]

names = ["df1", "df2", "df3"]

fs = fsspec.filesystem("memory")

for name, csv in zip(names, csvs):
    with fs.open(f"{name}.csv", "w") as file:
        file.write(dedent(csv).strip())

duckdb.register_filesystem(fs)

duckdb.sql("""
from read_csv_auto('memory://df[123].csv', filename=true, header=true, union_by_name=true)
""")

# ┌─────────┬───────────┬─────────┬─────────┬─────────┬────────────────┬───────────────────┐
# │  first  │   last    │   age   │  city   │ postal  │     phone      │     filename      │
# │ varchar │  varchar  │ varchar │ varchar │ varchar │    varchar     │      varchar      │
# ├─────────┼───────────┼─────────┼─────────┼─────────┼────────────────┼───────────────────┤
# │ Aaaaa   │ Fffff     │ 40      │ NULL    │ NULL    │ NULL           │ memory:///df1.csv │
# │ Mmmm    │ Cccccc    │ 59      │ NULL    │ NULL    │ NULL           │ memory:///df1.csv │
# │ Hhhhh   │ Mmmmmm    │ 39      │ NULL    │ NULL    │ NULL           │ memory:///df1.csv │
# │ Cccccc  │ Rrrrrr    │ 42      │ NULL    │ NULL    │ NULL           │ memory:///df1.csv │
# │ Sssss   │ Niiiii    │ 32      │ NULL    │ NULL    │ NULL           │ memory:///df1.csv │
# │ Cccccc  │ Ssssssss  │ NULL    │ Pppppp  │ 156846  │ NULL           │ memory:///df2.csv │
# │ Ffffff  │ Ssss      │ NULL    │ Ssssss  │ 684321  │ NULL           │ memory:///df2.csv │
# │ Aa      │ Mmmmmm    │ NULL    │ Ggggg   │ 158431  │ NULL           │ memory:///df2.csv │
# │ Bbbbbb  │ Ccccccc   │ NULL    │ Kkkkkkk │ 543154  │ NULL           │ memory:///df2.csv │
# │ Bbbbbb  │ Mmmmmmmmm │ NULL    │ Rrrrrrr │ 354849  │ NULL           │ memory:///df2.csv │
# │ Rrr     │ Rr        │ old     │ NULL    │ C1O 2I2 │ (123) 456-4913 │ memory:///df3.csv │
# │ Ggggg   │ Ppppp     │ young   │ NULL    │ J8Z 7S8 │ (456) 789-9433 │ memory:///df3.csv │
# │ Ffffff  │ Gggg      │ young   │ NULL    │ N7A 5N4 │ (789) 321-2187 │ memory:///df3.csv │
# │ Iii     │ Ppppp     │ mid     │ NULL    │ N6H 3J8 │ (123) 321-4999 │ memory:///df3.csv │
# │ kkkkk   │ Dddddd    │ old     │ NULL    │ R0I 4V1 │ (456) 654-1427 │ memory:///df3.csv │
# ├─────────┴───────────┴─────────┴─────────┴─────────┴────────────────┴───────────────────┤
# │ 15 rows                                                                      7 columns │
# └────────────────────────────────────────────────────────────────────────────────────────┘
```

**Comment by ritchie46** [comment]
It does incur an overhead yes. How big I don't know.

**Comment by xuJ14** [comment]
Vote for this feature!

**Comment by alexander-beedie** [comment]
This option ("diagonal_relaxed") will be available in the upcoming `0.19.8` ✌️
